xarray/core/combine.py

```diff
@@ -369,24 +371,197 @@ def _auto_concat(datasets, dim=None, data_vars='all', coords='different'):
 _CONCAT_DIM_DEFAULT = '__infer_concat_dim__'
 
 
+def _infer_concat_order_from_positions(datasets, concat_dims):
+
+    combined_ids = OrderedDict(_infer_tile_ids_from_nested_list(datasets, ()))
+
+    tile_id, ds = list(combined_ids.items())[0]
+    n_dims = len(tile_id)
+    if concat_dims == _CONCAT_DIM_DEFAULT or concat_dims is None:
+        concat_dims = [concat_dims]*n_dims
+    else:
+        if len(concat_dims) != n_dims:
+            raise ValueError("concat_dims has length " + str(len(concat_dims))
+                             + " but the datasets passed are nested in a " +
+                             str(n_dims) + "-dimensional structure")
```

Comment:
Nit: usually it's more readable to write/read these with format strings (either using `%` or `.format()`), e.g., 
```python
            raise ValueError("concat_dims has length %d " 
                             "but the datasets passed are nested in a "
                             "%d-dimensional structure" % (len(concat_dims), n_dims))
```